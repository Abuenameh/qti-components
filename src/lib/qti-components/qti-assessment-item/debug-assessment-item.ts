import { html, LitElement } from 'lit';

import { customElement, state } from 'lit/decorators.js';
import { consume, ContextConsumer } from '@lit-labs/context';
import { ItemContext, itemContext } from './qti-assessment-item.context';

@customElement('debug-assessment-item')
export class DebugAssessmentItem extends LitElement {
  // @consume({ context: itemContext, subscribe: true })
  // @state()
  // private _context: Item;
  // connectedCallback(): void {
  //   super.connectedCallback();
  //   // new ContextConsumer(
  //   //   this,
  //   //   itemContext,
  //   //   e => this.dispatchEvent(new CustomEvent('context-changed', { bubbles: true, composed: true, detail: e })),
  //   //   true
  //   // );
  // }
  // render() {
  //   return html` <pre>${JSON.stringify(this._context.variables, null, 2)}</pre> `;
  // }
}
