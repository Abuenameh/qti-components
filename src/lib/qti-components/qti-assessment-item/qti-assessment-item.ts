import { customElement, property } from 'lit/decorators.js';
import { html, LitElement } from 'lit';
import { OutcomeVariable } from '../qti-utilities/Variables';
import { ResponseVariable } from '../qti-utilities/Variables';
import { watch } from '../../decorators/watch';
import type { ResponseInteraction } from '../qti-utilities/ExpressionResult';
import type { Interaction } from '../qti-interaction/internal/interaction/interaction';
import type { InteractionChangedDetails, OutcomeChangedDetails } from '../qti-utilities/EventTypes';
import type { QtiFeedback } from '../qti-feedback/qti-feedback';
import type { QtiResponseProcessing } from '../qti-responseprocessing';
import type { VariableDeclaration } from '../qti-utilities/Variables';
import type QtiRegisterVariable from '../qti-utilities/events/qti-register-variable';

/**
 * @summary The qti-assessment-item element contains all the other QTI 3 item structures.
 * @documentation https://www.imsglobal.org/spec/qti/v3p0/impl#h.dltnnj87l0yj
 * @status stable
 * @since 4.0
 *
 * @dependency qti-feedback
 * @dependency qti-responseprocessing
 *
 * @slot - The default slot where all the other QTI 3 item structures go.
 *
 * @event qti-interaction-changed - Emitted when an interaction is changed.
 * @event qti-outcome-changed - Emitted when an outcome has changed.
 * @event qti-response-processing - Emitted when response-processing is called.
 *
 */
@customElement('qti-assessment-item')
export class QtiAssessmentItem extends LitElement {
  public variables: VariableDeclaration<string | string[]>[] = []; // made public for tests to access
  private feedbackElements: QtiFeedback[] = [];
  private interactionElements: Interaction[] = [];

  @property({ type: Boolean }) disabled: boolean;
  @property({ type: Boolean }) readonly: boolean;

  @property({ type: String }) title: string;
  @property({ type: String }) identifier: string;

  private _numAttempts = 0;

  override render() {
    return html`<slot></slot>`;
  }

  @watch('disabled', { waitUntilFirstUpdate: true })
  _handleDisabledChange = (_: boolean, disabled: boolean) => {
    this.interactionElements.forEach(ch => (ch.disabled = disabled));
  };

  @watch('readonly', { waitUntilFirstUpdate: true })
  _handleReadonlyChange = (_: boolean, readonly: boolean) =>
    this.interactionElements.forEach(ch => (ch.readonly = readonly));

  constructor() {
    super();
    // listen for events.
    this.addEventListener('qti-register-variable', this._registerVariable);
    this.addEventListener('qti-register-feedback', this.registerFeedbackElement);
    this.addEventListener('qti-register-interaction', this.registerInteractionElement);
    this.addEventListener('qti-outcome-changed', this.outcomeChanged);
    this.addEventListener('qti-interaction-response', this.interactionResponse);
  }

  firstUpdated(val): void {
    super.firstUpdated(val);
    this.dispatchEvent(
      new CustomEvent<{ identifier: string }>('qti-item-connected', {
        bubbles: true,
        composed: true,
        detail: this
      })
    );
  }

  public override disconnectedCallback() {
    // remove events listeners.
    this.removeEventListener('qti-register-variable', this._registerVariable);
    this.removeEventListener('qti-register-feedback', this.registerFeedbackElement);
    this.removeEventListener('qti-register-interaction', this.registerInteractionElement);
    this.removeEventListener('qti-outcome-changed', this.outcomeChanged);
    this.removeEventListener('qti-interaction-response', this.interactionResponse);
  }

  private _registerVariable(event: QtiRegisterVariable) {
    this.variables.push(event.detail.variable);
  }

  public showCorrectResponse() {
    const responseVariables = this.variables.filter(
      (vari: ResponseVariable | OutcomeVariable) => 'correctResponse' in vari && vari.correctResponse
    ) as ResponseVariable[];
    this.responses = responseVariables.map(cr => {
      return {
        responseIdentifier: cr.identifier,
        response: cr.correctResponse
      };
    });
  }

  public processResponse(): boolean {
    const responseProcessor = this.querySelector('qti-response-processing') as unknown as QtiResponseProcessing;
    if (!responseProcessor) {
      console.info('Client side response processing template not available');
      return false;
    }

    if (!responseProcessor.process) {
      console.info('Client side response webcomponents not available');
      return false;
    }

    responseProcessor.process();
    this._numAttempts++;
    this.dispatchEvent(new CustomEvent('qti-response-processing'));
    return true;
  }

  // check set response back for interactions
  // We can set the responses as a string or as an object in ResponseFormat
  set responses(myResponses: ResponseInteraction[]) {
    if (myResponses) {
      for (const response of myResponses) {
        const interaction: Interaction | undefined = this.interactionElements.find(
          i => i.getAttribute('response-identifier') === response.responseIdentifier
        );
        // If there is a responseVariable, set the value back into the responseVariable
        const responseVariable = this.getResponse(response.responseIdentifier);
        if (responseVariable) {
          responseVariable.value = response.response;
        }
        // Set the response in the interaction
        if (interaction) {
          interaction.response = response.response;
        }
      }
    }
  }

  resetInteractions() {
    this.interactionElements.forEach(interactionElement => interactionElement.reset());
  }

  // check all interactions contain valid responses
  public validateResponses(): boolean {
    let result = true;
    this.interactionElements.forEach(interactionElement => {
      if (!interactionElement.validate()) {
        result = false;
      }
    });
    return result;
  }

  public getVariable(identifier: string): VariableDeclaration<string | string[] | undefined> {
    switch (identifier) {
      case 'numAttempts':
        return {
          identifier: 'numAttempts',
          cardinality: 'single',
          baseType: 'integer',
          value: this._numAttempts.toString()
        };
        break;

      default:
        {
          const variable = this.variables.find(vr => vr.identifier === identifier);
          if (!variable) {
            console.warn(`Variable with identifier ${identifier} was not found`);
            return null;
          }
          return variable;
        }
        break;
    }
  }

  public getResponse(identifier: string): ResponseVariable | null {
    const variable = this.variables.find(vr => vr.identifier === identifier);
    const responseVariable = variable instanceof ResponseVariable ? variable : null;
    return responseVariable;
  }

  public getOutcome(identifier: string): OutcomeVariable | null {
    const variable = this.variables.find(vr => vr.identifier === identifier);
    if (variable instanceof OutcomeVariable) {
      return variable;
    }
    return null;
  }

  private registerFeedbackElement(e: CustomEvent<QtiFeedback>) {
    e.stopPropagation();
    this.feedbackElements.push(e.detail);
  }

  private registerInteractionElement(e: CustomEvent<null>) {
    e.stopPropagation();
    this.interactionElements.push(e.target as Interaction);
  }

  // interaction has fired an event to save response.
  private interactionResponse(event: CustomEvent<InteractionChangedDetails>) {
    const detail = event.detail;
    // Maybe we do not have a responseVariable declared, no problem should work without
    const responseVariable = this.getResponse(detail.responseIdentifier);
    if (responseVariable) {
      responseVariable.value = detail.response;
    }
    event.stopImmediatePropagation();

    const identifier = this.getAttribute('identifier');
    if (!identifier) {
      console.warn(`qti-assessment-item has no identifier specified`);
    }

    this.dispatchEvent(
      new CustomEvent<InteractionChangedDetails>('qti-interaction-changed', {
        bubbles: true,
        composed: true,
        detail: {
          item: identifier,
          responseIdentifier: detail.responseIdentifier,
          response: detail.response
        }
      })
    );
  }

  // checks the attributes of all feedback elements and shows/hides as appropriate.
  private outcomeChanged(event: CustomEvent<OutcomeChangedDetails>) {
    this.showFeedback(event);
  }

  private showFeedback(event: CustomEvent<OutcomeChangedDetails>) {
    this.feedbackElements.forEach(fe => {
      fe.checkShowFeedback(event.detail.outcomeIdentifier); // , event.detail.value);
    });
  }

  public setOutcomeValue(identifier: string, value: string) {
    let outcomeIdentifier: VariableDeclaration<any>;
    switch (identifier) {
      // https://qti-components.citolab.nl/?path=/docs/qti-tutorial--docs#listing-12-using-numattempts-to-set-the-completion-status
      // PK: let's fake a completionStatus variable for adaptive items
      case 'completionStatus':
        outcomeIdentifier = {
          identifier: 'completionStatus',
          cardinality: 'single',
          baseType: 'string',
          value: value
        };
        break;
      default:
        {
          outcomeIdentifier = this.getOutcome(identifier);
          if (!outcomeIdentifier) {
            console.warn(`Can not set qti-outcome-identifier: ${identifier}, it is not available`);
            return;
          }

          if (outcomeIdentifier.cardinality === 'single') {
            outcomeIdentifier.value = value;
          } else {
            (outcomeIdentifier.value as any[]).push(value);
          }
        }
        break;
    }
    this.dispatchEvent(
      new CustomEvent<OutcomeChangedDetails>('qti-outcome-changed', {
        bubbles: true,
        composed: true,
        detail: {
          item: this.identifier,
          outcomeIdentifier: identifier,
          value: outcomeIdentifier.value
        }
      })
    );
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'qti-assessment-item': QtiAssessmentItem;
  }
}
