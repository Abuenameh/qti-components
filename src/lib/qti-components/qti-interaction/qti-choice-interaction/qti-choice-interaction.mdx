import { Canvas, Meta } from '@storybook/blocks';

import * as StylesheetStories from './qti-choice-interaction.stories';

<Meta of={StylesheetStories} />

# qti-choice-interaction

The ChoiceInteraction.Type (qti-choice-interaction) interaction presents a collection of choices to the candidate. The candidate's task is to select one or more of the choices, up to a maximum of max-choices. The interaction is always initialized with no choices selected.

<Canvas of={StylesheetStories.Interaction} />

<Canvas of={StylesheetStories.Simple} />
