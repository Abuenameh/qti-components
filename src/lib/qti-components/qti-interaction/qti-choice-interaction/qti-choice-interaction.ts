import { customElement, property } from 'lit/decorators.js';
import { html } from 'lit';
import styles from './qti-choice-interaction.styles';
import type { CSSResultGroup } from 'lit';
import { Choices } from '../internal/choices/choices';

/**
 * @summary The ChoiceInteraction.Type (qti-choice-interaction) interaction presents a collection of choices to the candidate.
 * @documentation https://www.imsglobal.org/spec/qti/v3p0/impl#h.j9nu1oa1tu3b
 * @status stable
 * @since 6.0
 *
 * @event qti-register-interaction - emitted when the interaction wants to register itself
 * @event qti-interaction-response - emitted when the interaction changes
 *
 * @slot - The default slot where <qti-simple-choice> must be placed.
 * @slot prompt - slot where the prompt is placed.
 */

@customElement('qti-choice-interaction')
export class QtiChoiceInteraction extends Choices {
  static styles: CSSResultGroup = styles;

  /** orientation of choices */
  @property({ type: String })
  public orientation: 'horizontal' | 'vertical';

  render() {
    return html` <slot name="prompt"></slot><slot part="slot"></slot>`;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'qti-choice-interaction': QtiChoiceInteraction;
  }
}
