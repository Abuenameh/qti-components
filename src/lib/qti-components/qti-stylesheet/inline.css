qti-assessment-item {
  display: flex;
  flex-orientation: column;
}

qti-assessment-item qti-simple-choice {
  position: relative !important;
  margin: 20px !important;
  background: #ff0000;
  color: rgb(63, 63, 63) !important;
  font-size: 40px !important;
  font-weight: bold !important;
  line-height: 1.3em !important;
  border: 2px dashed rgb(172, 172, 172) !important;
  border-radius: 0px !important;
  font-weight: 800 !important;
  font-family: 'Arial', 'Lucida Sans Unicode', 'Lucida Grande', 'Lucida Sans', Arial, sans-serif;
}

qti-assessment-item button:active {
  transform: translate(0px, 5px);
  border-bottom: 10px solid;
}

qti-assessment-item qti-simple-choice[aria-checked='true'] {
  background-color: white !important;
  color: #ff0000 !important;
  /* animation: zoom 0.2s ease-in; */
}
