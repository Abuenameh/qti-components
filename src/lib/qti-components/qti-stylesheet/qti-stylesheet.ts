import { LitElement, PropertyValueMap } from 'lit';
export class QtiStylesheet extends LitElement {
  private styleLink: HTMLStyleElement | HTMLLinkElement;

  constructor() {
    super();
  }

  protected firstUpdated(_changedProperties: PropertyValueMap<any> | Map<PropertyKey, unknown>): void {
    super.firstUpdated(_changedProperties);

    const item = this.closest('qti-assessment-item');
    const link = this.getAttribute('href');

    if (link !== null) {
      const styles = document.createElement('link');
      styles.rel = 'stylesheet';
      styles.type = 'text/css';
      styles.media = 'screen';
      styles.href = link;
      item.appendChild(styles);
      this.styleLink = styles;
    }

    if (this.textContent !== null) {
      const styles = document.createElement('style');
      styles.media = 'screen';
      styles.textContent = this.textContent;
      item.appendChild(styles);
      this.styleLink = styles;
    }
  }

  override disconnectedCallback() {
    const item = this.closest('qti-assessment-item');
    if (this.styleLink && this.styleLink.parentElement === this) {
      try {
        item.removeChild(this.styleLink);
      } catch (error) {
        console.log('could not remove stylesheet');
      }
    }
  }
}

customElements.define('qti-stylesheet', QtiStylesheet);
