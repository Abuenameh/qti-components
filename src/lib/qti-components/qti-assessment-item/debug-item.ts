import { html, LitElement } from 'lit';

import { customElement, queryAssignedElements, queryAssignedNodes, state } from 'lit/decorators.js';
import { QtiAssessmentItem } from './qti-assessment-item';
import { register } from 'module';
import { VariableDeclaration } from '../qti-utilities/Variables';

@customElement('debug-item')
export class DebugItem extends LitElement {
  private _assessmentItemRef: QtiAssessmentItem | undefined;

  @state()
  private _items: Array<VariableDeclaration<string | string[]>> = [];

  @state()
  private _savedResponses: ReadonlyArray<VariableDeclaration<string | string[]>> = [];

  private _registerItem = (item: QtiAssessmentItem) => {
    this._assessmentItemRef = item;
    this._assessmentItemRef.context;
  };

  private _contextChanged = () => {
    this._items = [...(this._assessmentItemRef?.context.variables ?? [])];
  };

  private _saveResponse = () => {
    this._savedResponses = this._assessmentItemRef.context.variables;
  };

  private _loadResponse = () => {
    this._assessmentItemRef.context = { ...this._assessmentItemRef.context, variables: this._savedResponses };
  };
  render() {
    return html`
      <pre>${JSON.stringify(this._items, null, 2)}</pre>
      <button @click=${this._saveResponse}>Save</button>
      <button @click=${this._loadResponse}>Load</button>

      <slot
        @item-connected=${({ detail }) => this._registerItem(detail)}
        @context-changed=${() => this._contextChanged()}
      >
      </slot>
      <pre>${JSON.stringify(this._savedResponses, null, 2)}</pre>
    `;
  }
}

// @queryAssignedElements({ slot: 'list', flatten: true, selector: 'qti-assessment-item' })
// listItems!: Array<QtiAssessmentItem>;

// @slotchange=${(e: Event) => console.log('slotchange', e, this.listItems)}
