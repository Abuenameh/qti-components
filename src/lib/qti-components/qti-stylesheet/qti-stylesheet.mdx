import { Canvas, Meta, Story } from '@storybook/blocks';

import * as StylesheetStories from './qti-stylesheet.stories';

<Meta of={StylesheetStories} />

# stylesheet

A stylesheet is a square box that can be activated or deactivated when ticked.

Use stylesheets to select one or more options from a list of choices.

<Canvas of={StylesheetStories.Default} />
