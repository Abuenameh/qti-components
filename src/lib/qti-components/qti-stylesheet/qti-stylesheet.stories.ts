import { html } from 'lit-html';
import type { Meta, StoryObj } from '@storybook/web-components';

import '../qti-assessment-item/qti-assessment-item';
import '../qti-item-body/qti-item-body';
import '../qti-interaction/qti-choice-interaction/qti-choice-interaction';
import '../qti-interaction/internal/choice/qti-choice';
import '../qti-stylesheet/qti-stylesheet';
import { QtiStylesheet } from '../qti-stylesheet/qti-stylesheet';

const meta = {
  /* ðŸ‘‡ The title prop is optional.
   * See https://storybook.js.org/docs/7.0/react/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  component: 'qti-stylesheet'
} satisfies Meta<typeof QtiStylesheet>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => html`<qti-assessment-item>
    <qti-item-body>
      <qti-stylesheet href="/qti-stylesheet/rijmkonijn.css" type="text/css"></qti-stylesheet>
      <qti-choice-interaction
        class="qti-orientation-horizontal qti-input-control-hidden"
        max-choices="1"
        shuffle="false"
      >
        <qti-simple-choice identifier="A"> Boer </qti-simple-choice>
        <qti-simple-choice identifier="B"> Doek </qti-simple-choice>
        <qti-simple-choice identifier="C"> Voet </qti-simple-choice>
      </qti-choice-interaction>
    </qti-item-body>
  </qti-assessment-item>`
};
