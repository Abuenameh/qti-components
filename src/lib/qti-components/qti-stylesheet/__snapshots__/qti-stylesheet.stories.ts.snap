// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/qti-components/qti-stylesheet Default smoke-test 1`] = `
<div id="root-inner">
  <qti-assessment-item>
    <qti-stylesheet href="/qti-stylesheet/linked.css"
                    type="text/css"
    >
    </qti-stylesheet>
    <qti-item-body>
      <qti-choice-interaction class="qti-orientation-horizontal qti-input-control-hidden"
                              max-choices="1"
                              shuffle="false"
      >
        <qti-simple-choice identifier="A">
          Boer
        </qti-simple-choice>
        <qti-simple-choice identifier="B">
          Doek
        </qti-simple-choice>
        <qti-simple-choice identifier="C">
          Voet
        </qti-simple-choice>
      </qti-choice-interaction>
    </qti-item-body>
    <link rel="stylesheet"
          type="text/css"
          media="screen"
          href="/qti-stylesheet/linked.css"
    >
    <style media="screen">
    </style>
  </qti-assessment-item>
</div>
`;

exports[`lib/qti-components/qti-stylesheet Inline smoke-test 1`] = `
<div id="root-inner">
  <qti-assessment-item>
    <qti-stylesheet type="text/css">
      qti-assessment-item {
  display: flex;
  flex-orientation: column;
}

qti-assessment-item qti-simple-choice {
  position: relative !important;
  margin: 20px !important;
  background: #ff0000;
  color: rgb(63, 63, 63) !important;
  font-size: 40px !important;
  font-weight: bold !important;
  line-height: 1.3em !important;
  border: 2px dashed rgb(172, 172, 172) !important;
  border-radius: 0px !important;
  font-weight: 800 !important;
  font-family: 'Arial', 'Lucida Sans Unicode', 'Lucida Grande', 'Lucida Sans', Arial, sans-serif;
}

qti-assessment-item button:active {
  transform: translate(0px, 5px);
  border-bottom: 10px solid;
}

qti-assessment-item qti-simple-choice[aria-checked='true'] {
  background-color: white !important;
  color: #ff0000 !important;
  /* animation: zoom 0.2s ease-in; */
}
    </qti-stylesheet>
    <qti-item-body>
      <qti-choice-interaction class="qti-orientation-horizontal qti-input-control-hidden"
                              max-choices="1"
                              shuffle="false"
      >
        <qti-simple-choice identifier="A">
          Boer
        </qti-simple-choice>
        <qti-simple-choice identifier="B">
          Doek
        </qti-simple-choice>
        <qti-simple-choice identifier="C">
          Voet
        </qti-simple-choice>
      </qti-choice-interaction>
    </qti-item-body>
    <style media="screen">
      qti-assessment-item {
  display: flex;
  flex-orientation: column;
}

qti-assessment-item qti-simple-choice {
  position: relative !important;
  margin: 20px !important;
  background: #ff0000;
  color: rgb(63, 63, 63) !important;
  font-size: 40px !important;
  font-weight: bold !important;
  line-height: 1.3em !important;
  border: 2px dashed rgb(172, 172, 172) !important;
  border-radius: 0px !important;
  font-weight: 800 !important;
  font-family: 'Arial', 'Lucida Sans Unicode', 'Lucida Grande', 'Lucida Sans', Arial, sans-serif;
}

qti-assessment-item button:active {
  transform: translate(0px, 5px);
  border-bottom: 10px solid;
}

qti-assessment-item qti-simple-choice[aria-checked='true'] {
  background-color: white !important;
  color: #ff0000 !important;
  /* animation: zoom 0.2s ease-in; */
}
    </style>
  </qti-assessment-item>
</div>
`;
